### PR TITLE
fix(types): v2

### PR DESCRIPTION
**Pull request by wolverineks**
working on updating types for v2

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -380,12 +379,28 @@ export interface QueryResultBase<TResult, TError = Error> {
   isFetching: boolean
   isStale: boolean
   failureCount: number
+  canFetchMore: boolean
+  markedForGarbageCollection: boolean
+  query: Object
```
Comment (MichaelDeBoey) [inline types/index.d.ts:384]:
```suggestion
  query: object
```

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by youngbob** [comment]
Since you're already working on this - `MutateOptions` is missing `throwOnError`. Thanks!

**Comment by tannerlinsley** [comment]
How's it going over here?

**Comment by wolverineks** [comment]
> How's it going over here?

only `prefetch` and `setQueryData` remaining

**Comment by wolverineks** [comment]
i think this has everything except `queryCache.prefetchQueries`
thats gonna need fresher eyes that i have at the moment

**Comment by kizivat** [comment]
I wanted to do the `queryCache.prefetchQuery` but it doesn't seem I can add commits to someone else's PR. So suggested diff:

```diff
@@ -578,8 +612,7 @@ export interface QueryCache {
       | undefined
       | (() => TKey | false | null | undefined),
     queryFn: QueryFunction<TResult, TKey>,
-    config?: QueryOptions<TResult, TError>,
-    options?: PrefetchQueryOptions<TResult, TError>, 
+    config?: PrefetchQueryOptions<TResult, TError>
   ): Promise<TResult>
 
   prefetchQuery<TResult, TKey extends string, TError = Error>(
@@ -590,8 +623,7 @@ export interface QueryCache {
       | undefined
       | (() => TKey | false | null | undefined),
     queryFn: QueryFunction<TResult, [TKey]>,
-    config?: QueryOptions<TResult, TError>,
-    options?: PrefetchQueryOptions<TResult, TError>
+    config?: PrefetchQueryOptions<TResult, TError>
   ): Promise<TResult>
 
   prefetchQuery<TResult, TKey extends AnyQueryKey, TError = Error>(
@@ -602,8 +634,7 @@ export interface QueryCache {
       | undefined
       | (() => TKey | false | null | undefined),
     queryFn: QueryFunction<TResult, TKey>,
-    config?: QueryOptions<TResult, TError>,
-    options?: PrefetchQueryOptions<TResult, TError>
+    config?: PrefetchQueryOptions<TResult, TError>
   ): Promise<TResult>
 
   prefetchQuery<TResult, TKey extends string, TError = Error>(
@@ -614,12 +645,12 @@ export interface QueryCache {
       | undefined
       | (() => TKey | false | null | undefined),
     queryFn: QueryFunction<TResult, [TKey]>,
-    config?: QueryOptions<TResult, TError>,
-    options?: PrefetchQueryOptions<TResult, TError>
+    config?: PrefetchQueryOptions<TResult, TError>
   ): Promise<TResult>
 
   prefetchQuery<TResult, TKey extends AnyQueryKey, TError = Error>({
     queryKey,
+
     queryFn,
     config,
   }: {
@@ -630,8 +661,7 @@ export interface QueryCache {
       | undefined
       | (() => TKey | false | null | undefined)
     queryFn: QueryFunction<TResult, TKey>
-    config?: QueryOptions<TResult, TError>,
-    options?: PrefetchQueryOptions<TResult, TError>
+    config?: PrefetchQueryOptions<TResult, TError>
   }): Promise<TResult>
```

**Comment by kizivat** [comment]
I'll open other PR for it, feel free to include it in this one if you'd prefer to keep it under one roof

**Comment by wolverineks** [comment]
the types for prefetch are in another pr 
https://github.com/tannerlinsley/react-query/pull/633

**Comment by tannerlinsley** [comment]
You guys could try and use the new github code suggestions for those prefetch types? https://github.blog/2018-10-16-future-of-software/#suggested-changes-public-beta

**Comment by kizivat** [comment]
> You guys could try and use the new github code suggestions for those prefetch types? https://github.blog/2018-10-16-future-of-software/#suggested-changes-public-beta

To do so in this PR we'd need some change in somewhere near those lines of code as it doesn't seem to be possible to add comments/suggestions outside of a narrow context of the changed lines.

**Comment by tannerlinsley** [comment]
Darn!

**Comment by dcastil** [comment]
@kizivat You can also open a PR against `wolverineks:master` to submit your changes. When the PR gets merged, the changes will appear in this PR.

**Comment by wolverineks** [comment]
:grimacing: maaaaybe?

**Comment by kizivat** [comment]
So doesn't the `queryFn` still need to be optional as well?

**Comment by wolverineks** [comment]
> So doesn't the `queryFn` still need to be optional as well?

i think i misremembered that, it doesnt look like it is optional

**Comment by wolverineks** [comment]
> So doesn't the `queryFn` still need to be optional as well?

wellp, it is optional,and defaults to config.queryFn
not sure how to make that type safe...

**Comment by tannerlinsley** [comment]
Anything you guys are waiting on me for?

**Comment by wolverineks** [comment]
ok, dtslint passes

**Comment by tannerlinsley** [comment]
Cool , you should be able to add dtslint to the test:ci script in the package.json

**Comment by tannerlinsley** [comment]
Fantastic!

**Comment by wolverineks** [comment]
i wasnt sure about `infiniteQueryWithVariables`. this test looks obsolete, maybe?
https://github.com/tannerlinsley/react-query/blob/2767395101642d717af4fe8f1c40d1bbe5e5a03a/types/test.ts#L361

**Comment by tannerlinsley** [comment]
I think I’d still viable. The variables it is referring to are not the optional variables that were deprecated. I think. 😂

**Comment by wolverineks** [comment]
> I think I’d still viable. The variables it is referring to are not the optional variables that were deprecated. I think. 😂

could you explain a bit? i dont understand what this test is testing.

**Comment by tannerlinsley** [comment]
You're right. Useless. Get it out of there! ;)

**Comment by bugzpodder** [comment]
Is this missing a resetErrorBoundaries definition on QueryCache?

**Comment by tannerlinsley** [comment]
This is looking great. Do you all think these are good enough to release in their current state? Then we can just continue to improve and fix them as we go?

**Comment by bugzpodder** [comment]
LGTM

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by tannerlinsley** [comment]
Wow, thanks everyone here for working on this. I appreciate you. If you played any part in these great types, please feel free to open an issue and tag yourself using the [all-contributors bot](https://allcontributors.org/docs/en/bot/usage#all-contributors-add) to add yourself to the contributor list.

**Comment by Ethan-Arrowood** [comment]
Congrats on shipping type definitions! I know first hand it is no easy task to author definitions for a JS project 😄 great work!

**Comment by johnnyreilly** [comment]
Nice work all!

**Comment by selbekk** [comment]
Fantastic work - thanks for putting in the effort!

**Comment by reddevil22** [comment]
Thanks for the work, everyone. Definitely looking forward to testing out the new types

**Comment by kachkaev** [comment]
Many thanks for your work folks! 🎉 I was pretty impatient to try v2 with typings and installed 2.4.4 the first thing this morning 😅

All fits well in the tests project except one small thing in tests:

```ts
afterEach(() => {
  queryCache.clear({ notify: false }) // Expected 0 arguments, but got 1.ts(2554)
})
```

I can imagine a few more small hickups here and there, so wondering what’s the best place to track them. Does every possible error deserves its own PR or should we try collecting a few reports first and mitigate them as a batch?

**Comment by MichaelDeBoey** [comment]
@kachkaev I think you can just create a PR to fix that typing issue and we can fix them all when we come across them.
That way everybody can enjoy the fixed typings asap 🙂

**Pull request merged**